File: leetcode/501.find-mode-in-binary-search-tree/go/findmode.go
package findmodeinbst

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findMode(root *TreeNode) []int {
	var prev *TreeNode
	res := make([]int, 0)
	maxNum := 0
	currNum := 0

	var traversal func(*TreeNode)

	traversal = func(node *TreeNode) {
		if node == nil {
			return
		}

		traversal(node.Left)

		if prev != nil && prev.Val == node.Val {
			currNum++
		} else {
			currNum = 1
		}

		if currNum == maxNum {
			res = append(res, node.Val)
		} else if currNum > maxNum {
			maxNum = currNum
			res = []int{node.Val}
		}

		prev = node

		traversal(node.Right)
	}

	traversal(root)

	return res
}

File: leetcode/501.find-mode-in-binary-search-tree/ts/findmode.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const findMode = (root: TreeNode | null): number[] => {
    let res: number[] = []
    let prev: TreeNode | null = null
    let maxNum = 0
    let currNum = 0

    const traversal = (node: TreeNode | null) => {
        if (node === null) return

        traversal(node.left)

        if (prev === null || prev.val !== node.val) currNum = 1
        else currNum++

        if (currNum > maxNum) {
            maxNum = currNum
            res = [node.val]
        } else if (currNum === maxNum)
            res.push(node.val)

        prev = node
        traversal(node.right)
    }

    traversal(root)
    return res
}

export default findMode
